addons/mrp/report/mrp_report_bom_structure.py

```diff
@@ -310,32 +446,141 @@ def _get_sub_lines(self, bom, product_id, line_qty, line_id, level, unfolded_ids
+            if self.env.context.get('warehouse'):
+                domain.append(('warehouse_id', '=', self.env.context.get('warehouse')))
+
+            # Seek the closest date in the forecast report where consummed quantity >= forecasted quantity
+            closest_forecasted = self.env['report.stock.quantity']._read_group(domain, ['min_date:min(date)', 'product_id'], ['product_id'])
+            if closest_forecasted:
+                days_to_forecast = (closest_forecasted[0]['min_date'] - date_today).days
+                return ('expected', days_to_forecast)
+        return ('unavailable', False)
+
+    @api.model
+    def _get_resupply_availability(self, route_info, components):
+        if route_info.get('route_type') == 'manufacture':
+            max_component_delay = self._get_max_component_delay(components)
+            if max_component_delay is False:
+                return ('unavailable', False)
+            produce_delay = route_info.get('manufacture_delay', 0) + max_component_delay
+            return ('estimated', produce_delay)
+        return ('unavailable', False)
+
+    @api.model
+    def _get_max_component_delay(self, components):
+        max_component_delay = 0
+        for component in components:
+            line_delay = component.get('availability_delay', False)
+            if line_delay is False:
+                # This component isn't available right now and cannot be resupplied, so the manufactured product can't be resupplied either.
+                return False
+            max_component_delay = max(max_component_delay, line_delay)
+        return max_component_delay
 
-    def _get_extra_column_count(self):
-        return 0
+    @api.model
+    def _format_date_display(self, state, delay):
+        date_today = self.env.context.get('from_date', fields.date.today())
+        if state == 'available':
+            return _('Available')
+        if state == 'unavailable':
+            return _('Not Available')
+        if state == 'expected':
+            return _('Expected %s', format_date(self.env, date_today + timedelta(days=delay)))
+        if state == 'estimated':
+            return _('Estimated %s', format_date(self.env, date_today + timedelta(days=delay)))
+        return ''
+
+    @api.model
+    def _has_attachments(self, data):
+        return data['attachment_ids'] or any(self._has_attachments(component) for component in data.get('components', []))
+
+    @api.model
+    def get_warehouses(self):
+        return self.env['stock.warehouse'].search_read([('company_id', '=', self.env.company.id)], fields=['id', 'name'])
+
+    @api.model
+    def _get_resupply_rule(self, warehouse, product):
+        found_rules = product._get_rules_from_location(warehouse.lot_stock_id)
+        return found_rules[0] if found_rules else found_rules
```

Comment:
A bit overkill to have this function for 2 lines. Also you should return all the rules. And filter them on action to format the rule info to the user but use the sum to have the correct lead days